packages/core/src/component.test.ts

```diff
@@ -0,0 +1,311 @@
+            component.attributeChangedCallback("name1", "undefined", undefined)
+            component.attributeChangedCallback("name1", value, value)
+
+            expect(component.propChanged).toHaveBeenCalledTimes(0)
+        })
+        it("does not call attributeChangedCallback when values are not undefined & are not equal", () => {
+            class TestComponent extends Component {
+                propChanged = jest.fn()
+
+                attributeChangedCallback(name: string, oldValue: string, newValue: string) {
+                    super.attributeChangedCallback(name, oldValue, newValue)
+                }
+            }
+            const component = new TestComponent()
+
+            const name = "name1", value1 = "value1", value2 = "value2"
+            component.attributeChangedCallback(name, value1, value2)
+            expect(component.propChanged).toHaveBeenCalledWith(name, value1, value2)
+        })
+    })
+    describe("data layer", () => {
+        it("creates actions in bindActionCreators", () => {
+            const action = {type: "a"}
+            const actionCreators = {
+                a: jest.fn(() => action)
+            }
+            const store = createEmptyStore(null)
+            store.dispatch = jest.fn()
+            class TestComponent extends Component {
+                actionCreators = actionCreators
+                store = store
+                bindActionCreators() {
+                    super.bindActionCreators()
+                }
+                getActions() {
+                    return this.actions
+                }
+            }
+            const component = new TestComponent()
+
+            component.bindActionCreators()
+            const actionData = {b: 2}
+            component.getActions().a(actionData)
+
+            expect(store.dispatch).toHaveBeenCalledWith(action)
+            expect(actionCreators.a).toHaveBeenCalledWith(actionData)
+        })
+        it("updates state and calls callRender in update method", () => {
+            class TestComponent extends Component {
+                callRender = jest.fn()
+
+                callUpdate(state?: unknown) {
+                    this.update(state)
+                }
+                getState() {
+                    return this.state;
+                }
+            }
+            const component = new TestComponent()
+
+            component.callUpdate()
+            expect(component.getState()).toBe(undefined)
+
+            const newState = {a: 1}
+            component.callUpdate(newState)
+            expect(component.getState()).toBe(newState)
+
+            expect(component.callRender).toHaveBeenCalledTimes(2)
+        })
+        it("pass correct params to the subscribeWithSelector", () => {
+            resetDataLayer();
+
+            const selector = identity
+            const store = createEmptyStore(null)
+
+            class TestComponent extends Component {
```

Comment:
Probably you can make `TestComponent` out of the scope particular tests, because every time you create different scenario and methods for test.
Also, in case of intersections (like example -> state, selector) just make `reset()` method for the origin data restore. + I've been looking, and for the shallow review, I can't find intersections, test cases are independent

Reply:
it's for readability - overwritten code is near tests so you can easily view it and understand what's going on and what is overwritten and how

Reply:
I suppose you can make one class for all test scenario - it will remain readable

Reply:
no - it will have a lot of overwritten methods - it's not readable and clear when each method is needed